### DISCUSSION TITLE
[Inconsistent message] Adapt CS0236 message text to post C# 12.0

### DISCUSSION THREAD

**Discussion by zorgoz** [dotnet/csharplang#8559] [General]
The message currently looks like `CS0236 A field initializer cannot reference the non-static field, method, or property ....`
However, this is not true anymore:

 `Value` is indeed a non-static member, that might or might not be initialized at that point.
```c#
public class X1
{ 
	public int Value { get; }
	public int DoubleValue { get; set; } = Value*2; // << CS0236 on Value
	
	public X1(int value)
	{
		Value = value;
	}
}
```

However, in the following scenario, there is no error, as the primary constructor will initialize `Value` ahead of that statement.
But, the message does not include this scenario as an _exception_ to the rule.

```c#
public class X2(int Value)
{
	public int DoubleValue { get; set; } = Value * 2;
}
```

**Comment by jnm2** [comment]
A primary parameter is none of the things mentioned in the error message, so the error message is accurate. The error message doesn't list things that are valid; that would be a very long list.

**Reply by zorgoz** [reply]
The parameter is not, but the backing filed referenced by the expression is a non-static field.  Unless it is using the parameter directly and not the backing field.

**Reply by HaloFour** [reply]
Correct, it is using the parameter directly.  Also, since the parameter is not used again post-construction, no field is emitted for it.

**Reply by jnm2** [reply]
@zorgoz Even if there is a backing field in the lowering strategy, that would not matter from the lens of the language itself. It would be mixing abstraction layers, so to speak.

**Comment by TahirAhmadov** [comment]
https://github.com/dotnet/csharplang/discussions/6128
